pkg/edit/navigation_test.go

```diff
@@ -63,23 +66,91 @@ func TestNavigation_WidthRatio(t *testing.T) {
 	evals(f.Evaler, `@edit:navigation:width-ratio = 1 1 1`)
 	f.TTYCtrl.Inject(term.K('N', ui.Ctrl))
 	f.TestTTY(t,
-		"~"+string(os.PathSeparator)+"d> ", term.DotHere, "\n",
+		filepath.Join("~", "d"), "> ", term.DotHere, "\n",
 		" NAVIGATING  \n", Styles,
 		"************ ",
-		" d                a               ", Styles,
+		" d                a               \n", Styles,
 		"################ ++++++++++++++++ ",
+		"                  e              ", Styles,
+		"                 ////////////////",
+	)
+}
+
+// Test some corner cases:
+//
+// 1) Inserting a selection when the CLI cursor is at the start of the edit
+// buffer omits the space char prefix.
+//
+// 2) Inserting the "selection" in an empty directory inserts nothing rather
+// than the literal ` ''`  that used to be inserted.
+//
+// 3) Inserting a selection when the CLI cursor is not at the start of the
+// edit buffer but the preceding char is a space does not prepend another
+// space.
+//
+// See https://github.com/elves/elvish/issues/1169
+func TestNavigationCornerCases(t *testing.T) {
+	f, cleanup := setupNav()
+	defer cleanup()
+
+	// Validate that inserting a file name at the start of the command line
+	// does not prepend a space.
+	f.TTYCtrl.Inject(term.K('N', ui.Ctrl))
+	f.TTYCtrl.Inject(term.K(ui.Enter, ui.Alt))
+	f.TestTTY(t,
+		filepath.Join("~", "d"), "> ",
+		"a", Styles,
+		"!", term.DotHere, "\n",
+		" NAVIGATING  \n", Styles,
+		"************ ",
+		" d      a                 \n", Styles,
+		"###### ++++++++++++++++++ ",
+		"        e                ", Styles,
+		"       //////////////////",
+	)
+
+	// Validate that inserting an empty directory is a nop.
+	f.TTYCtrl.Inject(term.K(ui.Down))          // select "e"
+	f.TTYCtrl.Inject(term.K(ui.Right))         // move into the empty "e" directory
+	f.TTYCtrl.Inject(term.K(ui.Enter, ui.Alt)) // insert nothing since the dir is empty
+	f.TestTTY(t,
```

Comment:
This test here might pass even if pressing Enter inserts something like `''` - it could catch the state before that string is entered.

This is because the editor is running on a different goroutine, and TestTTY only tests that the expected state appears once, and doesn't guarantee that it is the steady state.

You can fix this by adding something after the Enter, for example typing a '#' and expect to see only a `#` in the code area.

Reply:
Good point. Fixed. However, I feel like I still don't understand why what you wrote is needed. Doesn't this apply to the other tests in this module? For clarity shouldn't there be a method we can call to force syncing the state of the two goroutines?

Reply:
The problem here is that you are testing the absence of change of the UI; typically UI tests test against the change of the UI.